### PR TITLE
Implement FixupPrecode for Arm64

### PR DESCRIPTION
**Pull request by ramarag**

### CODE REVIEW

**Comment by ramarag** [comment]
@dotnet-bot test Windows_NT arm64 Checked
@dotnet-bot test Windows_NT arm64 Release pri1r2r

**Comment by ramarag** [comment]
@dotnet-bot test Windows_NT arm64 Checked

**Comment by ramarag** [comment]
@dotnet-bot test Windows_NT arm64 Cross Release

**Comment by ramarag** [comment]
@dotnet-bot test Windows_NT arm64 Cross Release

**Comment by ramarag** [comment]
@dotnet-bot test Windows_NT arm64 Release pri1r2r

**Comment by ramarag** [comment]
@dotnet-bot test Windows_NT arm64 Checked

**Comment by ramarag** [comment]
@kyulee1 how do i debug the issue with the following failure:
http://dotnet-ci.cloudapp.net/job/dotnet_coreclr/job/master/job/arm64_cross_release_windows_nt_pri1r2r_prtest/17/consoleFull#-1176242260420e5adf-a1af-470a-bb97-8255d5af8765

**Comment by kyulee1** [comment]
Machines are offline and unreachable, which frequently occurred these days.
I took the affected machines out of the pool temporarily.

**Comment by kyulee1** [comment]
@dotnet-bot test Windows_NT arm64 Release pri1r2r

**Comment by ramarag** [comment]
@dotnet-bot test Windows_NT arm64 Release pri1r2r

**Comment by ramarag** [comment]
@dotnet-bot test Windows_NT arm64 Checked

**Comment by ramarag** [comment]
@dotnet-bot test Windows_NT arm64 Release pri1r2r

**Comment by ramarag** [comment]
@dotnet-bot test Windows_NT arm64 Checked

**Comment by ramarag** [comment]
@dotnet/arm64-contrib @rahku PTAL

File: src/vm/arm64/asmhelpers.asm [hunk 1]
```diff
@@ -193,9 +193,25 @@ Done
         NESTED_END
 
 ; ------------------------------------------------------------------
-; ARM64TODO: Implement PrecodeFixupThunk when PreCode is Enabled
+; The call in fixup precode initally points to this function.
+; The pupose of this function is to load the MethodDesc and forward the call the prestub.
         NESTED_ENTRY PrecodeFixupThunk
```
Comment (rahku) [inline src/vm/arm64/asmhelpers.asm]: forward the call *to* prestub

File: src/vm/arm64/cgencpu.h [hunk 2]
```diff
@@ -613,44 +613,76 @@ typedef DPTR(NDirectImportPrecode) PTR_NDirectImportPrecode;
...
     }
 
     BOOL SetTargetInterlocked(TADDR target, TADDR expected)
     {
-        _ASSERTE(!"ARM64:NYI");
-        return NULL;
+        CONTRACTL
+        {
+            THROWS;
+        GC_TRIGGERS;
```
Comment (rahku) [inline src/vm/arm64/cgencpu.h]: indentation

**Comment by rahku** [comment]
In precode.cpp in function Precode::SetTargetInterlocked
following code also applies for arm64. Can you make the change
#if !defined(_TARGET_ARM_)
    if (ret) {
        FlushInstructionCache(GetCurrentProcess(),this,SizeOf());
    }
#endif

File: src/vm/arm64/asmhelpers.asm [hunk 3]
```diff
@@ -193,9 +193,25 @@ Done
...
-        brk        #0
+
+        ; x12 = FixupPrecode *
+        ; On Exit
+        ; x12 = MethodDesc*
+        ; x13, x14 Trashed
+        ; Inline computation done by FixupPrecode::GetMethodDesc()
+        ldrb    w13, [x12, #Offset_PrecodeChunkIndex]              ; m_PrecodeChunkIndex
```
Comment (rahku) [inline src/vm/arm64/asmhelpers.asm]: Is this the disassembly of FixupPrecode::GetMethodDesc()

**Comment by rahku** [comment]
LGTM

File: src/vm/arm64/asmhelpers.asm (hunk 3 above)
Comment (ramarag) [inline src/vm/arm64/asmhelpers.asm]: NO the assembly for FixupPrecode::GetMethodDesc() generated by C++ compiler is actually 
39403c08 ldrb        w8,[x0,#0xF]
d280030a mov         x10,#0x18
11000508 add         w8,w8,#1
93407d09 sxtw        x9,w8
9b0a7d29 mul         x9,x9,x10
f860692a ldr         x10,[x9,x0]
b500006a cbnz        x10,coreclr!FixupPrecode::GetMethodDesc+0x24 (00000001`80145004)
d2800000 mov         x0,#0
14000004 b           coreclr!FixupPrecode::GetMethodDesc+0x30 (00000001`80145010)
39403808 ldrb        w8,[x0,#0xE]
531d7108 lsl         w8,w8,#3
8b28c140 add         x0,x10,w8 sxtw
d65f03c0 ret         lr

I replaced the mul with two add instructions. The assembly code here is similar to what is found in arm

**Comment by ramarag** [comment]
@dotnet-bot test Windows_NT arm64 Release pri1r2r

**Pull request merged**